File: Blind150/5.Binary Search/Python/search_in_2d_matrix.py

# https://leetcode.com/problems/search-a-2d-matrix/description/

from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROWS, COLS = len(matrix), len(matrix[0])

        t, b = 0, ROWS - 1

        row = None

        while t <= b:

            mid = t + (b-t) // 2

            if target > matrix[mid][-1]:

                t = mid + 1
            elif target < matrix[mid][0]:
                b = mid - 1
            else:
                row = mid
                break

        if row == None:
            return False

        l,r = 0 , COLS - 1

        while l <= r:

            mid = l + (r-l) // 2

            if matrix[row][mid] == target:
                return True

            elif matrix[row][mid] < target:
                l = mid + 1
            else:
                r = mid - 1

        return False

File: Blind150/5.Binary Search/Go/search_in_2d_matrix.go

// https://leetcode.com/problems/search-a-2d-matrix/description/
func searchMatrix(matrix [][]int, target int) bool {

    rows, cols := len(matrix), len(matrix[0])
    t,b := 0, rows - 1
    row := -1

    for t <= b{

        mid := t + (b -t) / 2

        if target > matrix[mid][cols - 1] {
            t = mid + 1
        }else if target < matrix[mid][0] {
            b = mid - 1
        }else{
            row = mid
            break
        }
    }

    if row == -1 {
        return false
    }

    l,r := 0, cols - 1

    for l <= r {
        rowMid := l + ( r-l) / 2

        if matrix[row][rowMid] == target{
            return true
        }else if target >  matrix[row][rowMid]{
            l = rowMid + 1
        }else{
            r = rowMid - 1
        }
    }
    return false

}
